File: libs/python/ggcommons/credentials/keyprovider.py
"""Key providers (KEK custodians). Phase 1 ships :class:`FileKeyProvider`.

The DEK is wrapped with AES-256-GCM under the KEK, AAD-bound to the vault id — identical to the
Rust reference, so a vault wrapped by one language unwraps in another.
"""
import base64
import os
from abc import ABC, abstractmethod

from . import crypto
from .errors import CredentialError
from .format import dek_wrap_aad


class KeyProvider(ABC):
    """Wraps/unwraps the vault DEK without exposing the KEK."""

    @property
    @abstractmethod
    def provider_id(self) -> str:
        ...

    @abstractmethod
    def wrap_dek(self, vault_id: str, dek: bytes) -> dict:
        """Return the ``kek`` dict persisted in the vault file."""

    @abstractmethod
    def unwrap_dek(self, vault_id: str, kek: dict) -> bytes:
        """Recover the DEK from a ``kek`` dict."""


class FileKeyProvider(KeyProvider):
    """KEK held as 32 bytes in a local key file (standalone / offline-fallback custodian)."""

    def __init__(self, kek: bytes):
        if len(kek) != crypto.KEY_LEN:
            raise CredentialError(f"KEK must be {crypto.KEY_LEN} bytes")
        self._kek = kek

    @classmethod
    def from_keyfile(cls, path: str) -> "FileKeyProvider":
        with open(path, "rb") as f:
            return cls(f.read())

    @classmethod
    def generate_keyfile(cls, path: str) -> "FileKeyProvider":
        kek = crypto.random(crypto.KEY_LEN)
        with open(path, "wb") as f:
            f.write(kek)
        try:
            os.chmod(path, 0o600)
        except OSError:
            pass
        return cls(kek)

    @property
    def provider_id(self) -> str:
        return "file"

    def wrap_dek(self, vault_id: str, dek: bytes) -> dict:
        nonce = crypto.random(crypto.NONCE_LEN)
        wrapped = crypto.seal(self._kek, nonce, dek_wrap_aad(vault_id), dek)
        return {
            "provider": "file",
            "alg": "AES-256-GCM",
            "wrapNonce": base64.b64encode(nonce).decode("ascii"),
            "wrappedDek": base64.b64encode(wrapped).decode("ascii"),
        }

    def unwrap_dek(self, vault_id: str, kek: dict) -> bytes:
        nonce_b = kek.get("wrapNonce")
        if not nonce_b:
            raise CredentialError("file KEK: missing wrapNonce")
        nonce = base64.b64decode(nonce_b)
        wrapped = base64.b64decode(kek["wrappedDek"])
        return crypto.open_(self._kek, nonce, dek_wrap_aad(vault_id), wrapped)


class KmsKeyProvider(KeyProvider):
    """KMS-wrapped DEK custodian: the DEK is encrypted by an AWS KMS CMK (the KEK never leaves KMS)
    and unwrapped via ``kms:Decrypt`` — using AWS creds / TES on Greengrass. The encryption context
    binds the wrapped DEK to the vault id (anti-swap). Mirrors the Rust ``mod kms``.

    ``boto3`` is imported lazily so non-KMS components don't require it at import time.
    """

    def __init__(self, key_id: str, region: str = None, endpoint_url: str = None):
        import boto3  # imported lazily so non-KMS components don't require it at import time

        if not key_id:
            raise CredentialError("kms key provider requires keyProvider.kmsKeyId")
        self._key_id = key_id
        self._client = boto3.client("kms", region_name=region, endpoint_url=endpoint_url)

    @property
    def provider_id(self) -> str:
        return "kms"

    def wrap_dek(self, vault_id: str, dek: bytes) -> dict:
        try:
            resp = self._client.encrypt(
                KeyId=self._key_id,
                Plaintext=dek,
                EncryptionContext={"vaultId": vault_id},
            )
        except Exception as e:
            raise CredentialError(f"kms encrypt: {e}") from None
        ct = resp.get("CiphertextBlob")
        if ct is None:
            raise CredentialError("kms encrypt: no ciphertext")
        return {
            "provider": "kms",
            "alg": "aws-kms",
            "wrappedDek": base64.b64encode(bytes(ct)).decode("ascii"),
            "kmsKeyId": self._key_id,
        }

    def unwrap_dek(self, vault_id: str, kek: dict) -> bytes:
        try:
            ct = base64.b64decode(kek["wrappedDek"])
        except Exception:
            raise CredentialError("kms: bad wrappedDek") from None
        try:
            resp = self._client.decrypt(
                CiphertextBlob=ct,
                KeyId=self._key_id,
                EncryptionContext={"vaultId": vault_id},
            )
        except Exception as e:
            raise CredentialError(f"kms decrypt: {e}") from None
        pt = resp.get("Plaintext")
        if pt is None:
            raise CredentialError("kms decrypt: no plaintext")
        return bytes(pt)


class Pkcs11KeyProvider(KeyProvider):
    """PKCS#11 (HSM/TPM/SoftHSM) DEK custodian — mirrors the Rust ``Pkcs11KeyProvider``. A
    non-extractable AES-256 key on the token is the KEK; the DEK is wrapped with AES-256-GCM
    *inside* the token, so the KEK never leaves hardware. The GCM AAD binds the wrapped DEK to the
    vault id (anti-swap) — same on-disk ``kek`` shape as :class:`FileKeyProvider` (provider
    ``pkcs11``, alg ``AES-256-GCM``, wrapNonce + wrappedDek).

    Uses ``python-pkcs11`` (imported lazily so non-HSM components don't require it). Selected via
    ``keyProvider.type = "pkcs11"`` with ``modulePath`` / ``tokenLabel`` / ``keyLabel`` and
    ``pinEnv`` (preferred) or ``pin``.
    """

    def __init__(self, module_path: str, token_label: str, key_label: str, pin: str):
        import pkcs11  # imported lazily so non-HSM components don't require it at import time

        self._key_label = key_label
        self._pin = pin
        try:
            self._lib = pkcs11.lib(module_path)
            self._token = self._lib.get_token(token_label=token_label)
        except Exception as e:
            raise CredentialError(f"pkcs11 load module/token '{token_label}': {e}") from None

    @property
    def provider_id(self) -> str:
        return "pkcs11"

    def _key(self, session):
        from pkcs11 import ObjectClass

        try:
            return session.get_key(object_class=ObjectClass.SECRET_KEY, label=self._key_label)
        except Exception as e:
            raise CredentialError(f"pkcs11 find key '{self._key_label}': {e}") from None

    def wrap_dek(self, vault_id: str, dek: bytes) -> dict:
        from pkcs11 import GCMParams, Mechanism

        iv = crypto.random(crypto.NONCE_LEN)
        aad = dek_wrap_aad(vault_id)
        try:
            with self._token.open(user_pin=self._pin) as session:
                key = self._key(session)
                ct = key.encrypt(dek, mechanism=Mechanism.AES_GCM, mechanism_param=GCMParams(iv, aad, 128))
        except CredentialError:
            raise
        except Exception as e:
            raise CredentialError(f"pkcs11 wrap: {e}") from None
        return {
            "provider": "pkcs11",
            "alg": "AES-256-GCM",
            "wrapNonce": base64.b64encode(iv).decode("ascii"),
            "wrappedDek": base64.b64encode(bytes(ct)).decode("ascii"),
        }

    def unwrap_dek(self, vault_id: str, kek: dict) -> bytes:
        from pkcs11 import GCMParams, Mechanism

        nonce_b = kek.get("wrapNonce")
        if not nonce_b:
            raise CredentialError("pkcs11 KEK: missing wrapNonce")
        iv = base64.b64decode(nonce_b)
        ct = base64.b64decode(kek["wrappedDek"])
        aad = dek_wrap_aad(vault_id)
        try:
            with self._token.open(user_pin=self._pin) as session:
                key = self._key(session)
                pt = key.decrypt(ct, mechanism=Mechanism.AES_GCM, mechanism_param=GCMParams(iv, aad, 128))
        except CredentialError:
            raise
        except Exception as e:
            raise CredentialError(f"pkcs11 unwrap: {e}") from None
        return bytes(pt)

File: libs/ts/src/credentials/keyprovider.ts
/** Key providers (KEK custodians). Phase 1 ships {@link FileKeyProvider}. */
import { readFileSync, writeFileSync } from "fs";

import { KEY_LEN, NONCE_LEN, open, random, seal } from "./crypto";
import { CredentialError } from "./errors";
import { dekWrapAad, KekInfo } from "./format";

/** Wraps/unwraps the vault DEK without exposing the KEK. */
export interface KeyProvider {
  providerId(): string;
  wrapDek(vaultId: string, dek: Buffer): KekInfo;
  unwrapDek(vaultId: string, kek: KekInfo): Buffer;
}

/** KEK held as 32 bytes in a local key file (standalone / offline-fallback custodian). */
export class FileKeyProvider implements KeyProvider {
  private readonly kek: Buffer;

  constructor(kek: Buffer) {
    if (kek.length !== KEY_LEN) {
      throw new CredentialError(`KEK must be ${KEY_LEN} bytes`);
    }
    this.kek = Buffer.from(kek);
  }

  static fromKeyFile(path: string): FileKeyProvider {
    return new FileKeyProvider(readFileSync(path));
  }

  static generateKeyFile(path: string): FileKeyProvider {
    const kek = random(KEY_LEN);
    writeFileSync(path, kek, { mode: 0o600 });
    return new FileKeyProvider(kek);
  }

  providerId(): string {
    return "file";
  }

  wrapDek(vaultId: string, dek: Buffer): KekInfo {
    const nonce = random(NONCE_LEN);
    const wrapped = seal(this.kek, nonce, dekWrapAad(vaultId), dek);
    return {
      provider: "file",
      alg: "AES-256-GCM",
      wrapNonce: nonce.toString("base64"),
      wrappedDek: wrapped.toString("base64"),
    };
  }

  unwrapDek(vaultId: string, kek: KekInfo): Buffer {
    if (!kek.wrapNonce) {
      throw new CredentialError("file KEK: missing wrapNonce");
    }
    const nonce = Buffer.from(kek.wrapNonce, "base64");
    const wrapped = Buffer.from(kek.wrappedDek, "base64");
    return open(this.kek, nonce, dekWrapAad(vaultId), wrapped);
  }
}

/**
 * KEK sourced as a RAW 32-byte key, base64-encoded, from an environment variable (typically a mounted
 * Kubernetes Secret) — the offline-capable software-KEK and the DEFAULT vault custodian on the
 * KUBERNETES platform (FR-CRED-3/FR-CRED-6). Cryptographically IDENTICAL to {@link FileKeyProvider}
 * given the same raw 32-byte KEK: it delegates wrap/unwrap to an internal {@link FileKeyProvider}
 * (same AES-256-GCM, same {@link dekWrapAad} AAD), so a vault wrapped by this provider with KEK `K` is
 * byte-compatible with one wrapped by `FileKeyProvider` with the same `K`. The ONLY differences are
 * {@link providerId} (`"env"`) and the {@link KekInfo} `provider` tag it writes. The KEK never touches
 * disk — it is read from the env var (a mounted Secret), not a key file.
 */
export class EnvKeyProvider implements KeyProvider {
  /** Delegate that owns the identical AES-256-GCM DEK wrap/unwrap crypto. */
  private readonly delegate: FileKeyProvider;

  /** Wrap a raw 32-byte KEK; length is validated by the delegate {@link FileKeyProvider}. */
  constructor(kek: Buffer) {
    this.delegate = new FileKeyProvider(kek);
  }

  /**
   * Read the base64-encoded KEK from environment variable `envVar`, base64-decode it, and validate it
   * is EXACTLY {@link KEY_LEN} (32) bytes.
   *
   * @throws {@link CredentialError} if the env var is unset/empty, not valid base64, or the decoded
   *         key is not 32 bytes.
   */
  static fromEnv(envVar: string): EnvKeyProvider {
    const rawValue = process.env[envVar];
    if (rawValue === undefined || rawValue === "") {
      throw new CredentialError(`env key provider: environment variable '${envVar}' is unset or empty`);
    }
    // Tolerate surrounding whitespace / a trailing newline — common when the value is sourced from a
    // mounted file / Secret (echo|base64, kubectl --from-file). Matches canonical Java (b64.trim()) and
    // Rust (raw.trim()) so the same Secret decodes identically across all four languages.
    const b64 = rawValue.trim();
    // Strict standard base64 (no embedded whitespace) — mirrors Java's basic Base64 decoder. Node's
    // Buffer.from(.,"base64") is lenient and never throws, so validate the shape explicitly first.
    if (b64.length % 4 !== 0 || !/^[A-Za-z0-9+/]*={0,2}$/.test(b64)) {
      throw new CredentialError(`env key provider: environment variable '${envVar}' is not valid base64`);
    }
    const kek = Buffer.from(b64, "base64");
    if (kek.length !== KEY_LEN) {
      throw new CredentialError(
        `env key provider: decoded KEK from '${envVar}' must be ${KEY_LEN} bytes, got ${kek.length}`,
      );
    }
    return new EnvKeyProvider(kek);
  }

  providerId(): string {
    return "env";
  }

  /** Wrap the DEK exactly as {@link FileKeyProvider} does, tagging the {@link KekInfo} `provider:"env"`. */
  wrapDek(vaultId: string, dek: Buffer): KekInfo {
    return { ...this.delegate.wrapDek(vaultId, dek), provider: "env" };
  }

  /** Unwrap the DEK via the delegate (the `provider` tag is irrelevant to the crypto). */
  unwrapDek(vaultId: string, kek: KekInfo): Buffer {
    return this.delegate.unwrapDek(vaultId, kek);
  }
}

/**
 * KMS-wrapped DEK custodian (mirrors the Rust `kms` module): the DEK is encrypted by an AWS KMS CMK
 * (the KEK never leaves KMS) and unwrapped via `kms:Decrypt` — using the default AWS credential chain
 * / TES on Greengrass. The encryption context `{vaultId}` binds the wrapped DEK to the vault id
 * (anti-swap). On-disk this produces `KekInfo{provider:"kms", alg:"aws-kms", wrappedDek:base64(ct),
 * kmsKeyId}` with no `wrapNonce`.
 *
 * ## Sync-vs-async approach
 * `KeyProvider.wrapDek`/`unwrapDek` are synchronous because `LocalVault.open` calls them inline, but
 * KMS calls are async. We resolve the round trip eagerly in {@link config!openFromConfig} (async),
 * then hand `LocalVault.open` a {@link PrewrappedKeyProvider} whose sync methods just return the
 * precomputed values. This keeps `vault.open` and the cross-language on-disk format unchanged.
 */
export class KmsKeyProvider {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private constructor(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private readonly client: any,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private readonly EncryptCommand: any,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private readonly DecryptCommand: any,
    private readonly keyId: string,
  ) {}

  /** Load `@aws-sdk/client-kms` (dynamically, so non-KMS components don't pull it) and bind a CMK. */
  static async create(keyId: string, region?: string, endpointUrl?: string): Promise<KmsKeyProvider> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let mod: any;
    try {
      mod = await import("@aws-sdk/client-kms");
    } catch {
      throw new CredentialError("kms key provider requires the @aws-sdk/client-kms package");
    }
    const client = new mod.KMSClient({ region, endpoint: endpointUrl });
    return new KmsKeyProvider(client, mod.EncryptCommand, mod.DecryptCommand, keyId);
  }

  providerId(): string {
    return "kms";
  }

  /** KMS-encrypt `dek` under the CMK, binding it to `vaultId` via the encryption context. */
  async wrapDek(vaultId: string, dek: Buffer): Promise<KekInfo> {
    let resp;
    try {
      resp = await this.client.send(
        new this.EncryptCommand({
          KeyId: this.keyId,
          Plaintext: dek,
          EncryptionContext: { vaultId },
        }),
      );
    } catch (e) {
      throw new CredentialError(`kms encrypt: ${(e as Error)?.message ?? String(e)}`);
    }
    if (!resp.CiphertextBlob) {
      throw new CredentialError("kms encrypt: no ciphertext");
    }
    return {
      provider: "kms",
      alg: "aws-kms",
      wrappedDek: Buffer.from(resp.CiphertextBlob).toString("base64"),
      kmsKeyId: this.keyId,
    };
  }

  /** KMS-decrypt the wrapped DEK described by `kek`, asserting the `vaultId` encryption context. */
  async unwrapDek(vaultId: string, kek: KekInfo): Promise<Buffer> {
    const ct = Buffer.from(kek.wrappedDek, "base64");
    let resp;
    try {
      resp = await this.client.send(
        new this.DecryptCommand({
          CiphertextBlob: ct,
          KeyId: this.keyId,
          EncryptionContext: { vaultId },
        }),
      );
    } catch (e) {
      throw new CredentialError(`kms decrypt: ${(e as Error)?.message ?? String(e)}`);
    }
    if (!resp.Plaintext) {
      throw new CredentialError("kms decrypt: no plaintext");
    }
    const pt = Buffer.from(resp.Plaintext);
    if (pt.length !== KEY_LEN) {
      throw new CredentialError("kms: unwrapped DEK wrong length");
    }
    return pt;
  }
}

/**
 * In-memory {@link KeyProvider} shim returning a pre-resolved KEK/DEK. Used to bridge the async KMS
 * round trip into `LocalVault.open`'s synchronous `wrapDek`/`unwrapDek` (see {@link KmsKeyProvider}).
 */
export class PrewrappedKeyProvider implements KeyProvider {
  constructor(
    private readonly id: string,
    private readonly kek: KekInfo,
    private readonly dek: Buffer,
  ) {}

  providerId(): string {
    return this.id;
  }

  wrapDek(): KekInfo {
    return this.kek;
  }

  unwrapDek(): Buffer {
    return Buffer.from(this.dek);
  }
}

/**
 * PKCS#11 (HSM/TPM/SoftHSM) DEK custodian — mirrors the Rust `Pkcs11KeyProvider`. A non-extractable
 * AES-256 key on the token is the KEK; the DEK is wrapped with AES-256-GCM *inside* the token (so the
 * KEK never leaves hardware). The GCM AAD binds the wrapped DEK to the vault id (anti-swap), so the
 * on-disk {@link KekInfo} shape is identical to {@link FileKeyProvider} (provider `"pkcs11"`, alg
 * `"AES-256-GCM"`, wrapNonce + wrappedDek).
 *
 * `graphene-pk11` (over `pkcs11js`) is synchronous, so `wrapDek`/`unwrapDek` stay synchronous and the
 * provider plugs straight into `LocalVault.open` — no async shim needed (unlike {@link KmsKeyProvider}).
 * Only the module load is async, done in {@link Pkcs11KeyProvider.create}. The binding is loaded
 * dynamically + listed in optionalDependencies, so non-HSM components don't pull the native addon.
 */
export class Pkcs11KeyProvider implements KeyProvider {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private constructor(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private readonly g: any,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private readonly session: any,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private readonly key: any,
  ) {}

  /** Load the module, select the slot whose token has `tokenLabel`, log in, and bind `keyLabel`. */
  static async create(
    modulePath: string,
    tokenLabel: string,
    keyLabel: string,
    pin: string,
  ): Promise<Pkcs11KeyProvider> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let mod: any;
    // Non-literal specifier so tsc treats this as a dynamic `any` import and does not require the
    // optional native package's types at compile time (graphene-pk11 is an optionalDependency).
    const pkg = "graphene-pk11";
    try {
      mod = await import(pkg);
    } catch {
      throw new CredentialError("pkcs11 key provider requires the graphene-pk11 package");
    }
    const g = mod.default ?? mod;
    const module = g.Module.load(modulePath, "edgecommons-vault");
    try {
      module.initialize();
    } catch (e) {
      // CKR_CRYPTOKI_ALREADY_INITIALIZED — another provider already initialized this module.
      if (!/ALREADY_INITIALIZED|0x191|already/i.test(String((e as Error)?.message ?? e))) {
        throw new CredentialError(`pkcs11 initialize: ${(e as Error)?.message ?? String(e)}`);
      }
    }
    const slots = module.getSlots(true);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let slot: any;
    for (let i = 0; i < slots.length; i++) {
      const s = slots.items(i);
      try {
        if (s.getToken().label.trim() === tokenLabel) {
          slot = s;
          break;
        }
      } catch {
        // Uninitialized / unreadable slot — skip.
      }
    }
    if (!slot) {
      throw new CredentialError(`pkcs11: no token labelled '${tokenLabel}' (module ${modulePath})`);
    }
    const session = slot.open(g.SessionFlag.SERIAL_SESSION | g.SessionFlag.RW_SESSION);
    try {
      session.login(pin);
    } catch (e) {
      // CKR_USER_ALREADY_LOGGED_IN — login state is per-token across sessions; another provider in
      // this process already logged in, which is fine.
      if (!/ALREADY_LOGGED_IN|0x100|already logged/i.test(String((e as Error)?.message ?? e))) {
        throw new CredentialError(`pkcs11 login: ${(e as Error)?.message ?? String(e)}`);
      }
    }
    const found = session.find({ class: g.ObjectClass.SECRET_KEY, label: keyLabel });
    if (found.length === 0) {
      throw new CredentialError(`pkcs11: no key labelled '${keyLabel}'`);
    }
    return new Pkcs11KeyProvider(g, session, found.items(0).toType());
  }

  providerId(): string {
    return "pkcs11";
  }

  wrapDek(vaultId: string, dek: Buffer): KekInfo {
    const iv = random(NONCE_LEN);
    const alg = { name: "AES_GCM", params: new this.g.AesGcmParams(iv, dekWrapAad(vaultId), 128) };
    let ct: Buffer;
    try {
      ct = Buffer.from(this.session.createCipher(alg, this.key).once(dek, Buffer.alloc(dek.length + 16)));
    } catch (e) {
      throw new CredentialError(`pkcs11 wrap: ${(e as Error)?.message ?? String(e)}`);
    }
    return {
      provider: "pkcs11",
      alg: "AES-256-GCM",
      wrapNonce: iv.toString("base64"),
      wrappedDek: ct.toString("base64"),
    };
  }

  unwrapDek(vaultId: string, kek: KekInfo): Buffer {
    if (!kek.wrapNonce) {
      throw new CredentialError("pkcs11 KEK: missing wrapNonce");
    }
    const iv = Buffer.from(kek.wrapNonce, "base64");
    const ct = Buffer.from(kek.wrappedDek, "base64");
    const alg = { name: "AES_GCM", params: new this.g.AesGcmParams(iv, dekWrapAad(vaultId), 128) };
    try {
      return Buffer.from(this.session.createDecipher(alg, this.key).once(ct, Buffer.alloc(ct.length)));
    } catch (e) {
      throw new CredentialError(`pkcs11 unwrap: ${(e as Error)?.message ?? String(e)}`);
    }
  }
}
